src/cashweb/wallet/index.ts

```diff
@@ -235,53 +210,72 @@ export class Wallet {
+    const tx = await chronikClient.tx(txid)
+    const txScripts = new Set<string>()
+    tx.inputs.forEach(
+      input => input.outputScript && txScripts.add(input.outputScript),
+    )
+    tx.outputs.forEach(output => txScripts.add(output.outputScript))
+    for (const scriptHex of txScripts) {
+      const script = new Script(scriptHex)
+      if (!script.isPublicKeyHashOut()) {
+        continue
+      }
+      const pkh = script.getPublicKeyHash().toString('hex')
+      if (!this.addressDataByPkh.has(pkh)) {
+        continue
+      }
+      await this.updateUTXOsFromPkh(pkh)
+    }
+  }
+
+  async updateUTXOsFromPkh(pkh: string) {
     const chronikClient = this.chronikClient
     assert(chronikClient, 'missing client in updateUTXOsFromScriptHash')
     try {
-      const { address, privKey } =
-        this.getAddressByElectrumScriptHash(scriptHash)
-      const chronikUtxos = await chronikClient
-        .script('p2pkh', privKey.toAddress().hashBuffer.toString('hex'))
-        .utxos()
+      const addressData = this.addressDataByPkh.get(pkh)
+      assert(addressData, `missing addressData for pkh ${pkh}`)
+      const chronikUtxos = await chronikClient.script('p2pkh', pkh).utxos()
       const utxos: Utxo[] = chronikUtxos.flatMap(scriptUtxos => {
         return scriptUtxos.utxos.map(utxo => ({
           txId: utxo.outpoint.txid,
           outputIndex: utxo.outpoint.outIdx,
           satoshis: utxo.value.toNumber(),
           type: 'p2pkh',
-          address,
-          privKey,
+          address: addressData.address,
+          privKey: addressData.privKey,
         }))
       })
-      await this.refreshUTXOsByAddr({ address, utxos })
+      await this.refreshUTXOsByAddr({ address: addressData.address, utxos })
     } catch (err) {
-      console.error('error deserializing utxo address', err, scriptHash)
+      console.error('error in updateUTXOsFromPkh', err, pkh)
     }
   }
 
   async updateHDUTXOs() {
     // Check HD Wallet
-    const scriptHashes = Object.keys(this.electrumScriptHashes)
     await P.map(
-      scriptHashes,
-      scriptHash => this.updateUTXOsFromScriptHash(scriptHash),
+      this.addressDataByPkh.keys(),
+      pkh => this.updateUTXOsFromPkh(pkh),
       { concurrency: 5 },
     )
   }
 
   async startListeners() {
-    const client = await this.electrumClientPromise
-    assert(client, 'missing client in startListeners')
-    const scriptHashes = Object.keys(this.electrumScriptHashes)
+    const chronikWs = this.chronikWs
+    assert(chronikWs, 'missing WebSocket client in startListeners')
+
+    chronikWs.onMessage = msg => this.addressUpdated(msg)
 
     await P.map(
-      scriptHashes,
-      scriptHash =>
-        client.subscribe(
-          this.scriptHashSubscriber,
-          'blockchain.scripthash.subscribe',
-          scriptHash,
+      [...this.walletKeys, ...this.changeKeys],
+      key =>
+        chronikWs.subscribe(
```

Comment:
On disconnect the state on the backend for listeners is lost. Does the chronik client resubscribe on it's own? The electrum client instrumentation would basically set the electrumClientPromise a second time which would trigger these functions again.

Reply:
Yes, chronik client keeps track of the subscriptions and re-subscribes on reconnect